File: 1012.rb
# You will get TLE with ruby in this problem. :)
class Student
    attr_accessor :id
    attr_accessor :s
    def initialize(a)
        @id = a[0]
        @s = [ a[1].to_i + a[2].to_i + a[3].to_i, a[1].to_i, a[2].to_i, a[3].to_i ]
    end
end
op = [ "A", "C", "M", "E" ]
n, m = gets.chomp.split(" ")
n = n.to_i; m = m.to_i
all = []
n.times { all << Student.new(gets.chomp.split(" ")) }
m.times do
    now = gets.chomp
    find = false
    s = []
    all.each { |x|
        if x.id == now
            find = true
            s = x.s
            break
        end
    }
    if !find
        puts "N/A"
        next
    end
    rank = [ 1, 1, 1, 1 ]
    all.each { |x|
        4.times { |i| rank[i] += 1 if x.s[i] > s[i] }
    }
    res = 2005; sel = nil
    4.times { |i|
        if rank[i] < res
            res = rank[i]
            sel = op[i]
        end
    }
    print res, " ", sel, "\n"
end

File: 1012.cpp
#include <string>
#include <cstring>
#include <iostream>
using namespace std;
string id[2005];
string op[4] = { "A", "C", "M", "E" };
int s[2005][4];
int main()
{
	int i, j, n, m;
	cin >> n >> m;
	for (i = 0; i < n; ++i) {
		cin >> id[i] >> s[i][1] >> s[i][2] >> s[i][3];
		s[i][0] = s[i][1] + s[i][2] + s[i][3];
	}
	for (j = 0; j < m; ++j) {
		string now;
		cin >> now;
		int score[4];
		bool flag = false;
		for (i = 0; i < n; ++i)
			if (id[i] == now)
			{
				memcpy(score, s[i], sizeof(s[i]));
				flag = true;
				break;
			}
		if (!flag)
		{
			cout << "N/A" << endl;
			continue;
		}
		int rank[4] = { 1, 1, 1, 1 };
		for (i = 0; i < n; ++i)
			for (int k = 0; k < 4; ++k)
				if (s[i][k] > score[k])
					++rank[k];
		int res = 2005;
		string sel = "";
		for (int k = 0; k < 4; ++k)
			if (res > rank[k])
			{
				res = rank[k];
				sel = op[k];
			}
		cout << res << " " << sel << endl;
	}
	return 0;
}

